DOCSP-39710 specify fields return (#56)
mongodb/docs-java-rs: source/specify-fields-return.txt

.. _java-rs-project:

========================
Specify Fields To Return
========================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: read, filter, project, select, choose

Overview
--------

> In this guide, you can learn how to specify which fields to return from a read
> operation by using a **projection**. A projection is a document that specifies
> which fields MongoDB returns from a query.

Reviewer: ```suggestion
In this guide, you can learn how to use the {+driver short+} to specify which fields to 
return from a read operation. You can select these fields by using a **projection**, a document that specifies
which fields MongoDB returns from a query.
```

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``sample_restaurants.restaurants`` collection
from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:ref:`<java-rs-getting-started>` guide.

.. include:: /includes/reactor-note.rst

Projection Types
----------------

> You can use a projection to specify which fields to include in a return
> document, or to specify which fields to exclude. You cannot combine inclusion and
> exclusion statements in a single projection, unless you are excluding the
> ``_id`` field.

Reviewer: ```suggestion
You can use a projection to specify which fields to include and exclude in retrieved
documents. By default, including certain fields excludes all other fields, so you cannot
combine inclusion and exclusion statements in a single projection,
unless you are excluding the ``_id`` field from results.
```

Specify Fields to Include
~~~~~~~~~~~~~~~~~~~~~~~~~

Use the following syntax to specify the fields to include in the result:

.. code-block:: java

   projection(fields(include("<field name>")))

> To use a projection, pass your query filter to the ``find()`` method. Then, chain the ``projection()`` method to the ``find()`` method. 
> The following example uses the ``find()`` method to find all restaurants with
> the ``name`` field value of ``"Emerald Pub"``. It then uses a projection to
> include only the ``name``, ``cuisine``, and ``borough`` fields from the returned documents.

Reviewer: ```suggestion
To use a projection, pass a query filter to the ``find()`` method. Then, chain the
``projection()`` method to the ``find()`` method call.
 
The following example uses the ``find()`` method to retrieve documents in
the ``name`` field value is ``"Emerald Pub"``. It then uses a projection to
include only the ``name``, ``cuisine``, and ``borough`` fields from the returned documents.
```

.. io-code-block::
    
   .. input:: /includes/project.java
      :start-after: start-project-include
      :end-before: end-project-include
      :language: java

   .. output::

      {'_id': ObjectId('...'), 'borough': 'Manhattan', 'cuisine': 'American', 'name': 'Emerald Pub'}
      {'_id': ObjectId('...'), 'borough': 'Queens', 'cuisine': 'American', 'name': 'Emerald Pub'}

When you use a projection to specify fields to include in the return
document, the ``_id`` field is also included by default. All other fields are

> implicitly excluded. To remove the ``_id`` field from the return
> document, you must :ref:`explicitly exclude it <java-rs-project-remove-id>`.

Reviewer: ```suggestion
implicitly excluded. You must :ref:`explicitly exclude the _id field <java-rs-project-remove-id>`
to omit it from the returned documents.
```

.. _java-rs-project-remove-id:

Exclude the ``_id`` Field
~~~~~~~~~~~~~~~~~~~~~~~~~

> When specifying fields to include, you can also exclude the ``_id`` field from
> the returned document.

Reviewer: ```suggestion
You can exclude the ``_id`` field from returned documents by using the ``excludeId()`` method.
```

> The following example performs the same query as the preceding example, but
> excludes the ``_id`` field from the projection:

Reviewer: ```suggestion
The following example performs the same query as in the preceding example, but
also specifies exclusion of the ``_id`` field in the returned documents:
```

Author: I kept it as "also excludes"; specifies exclusion feels very formal to me/a little awkward.

.. io-code-block::
    
   .. input:: /includes/project.java
      :start-after: start-project-include-without-id
      :end-before: end-project-include-without-id
      :language: java

   .. output::

      {'borough': 'Manhattan', 'cuisine': 'American', 'name': 'Emerald Pub'}
      {'borough': 'Queens', 'cuisine': 'American', 'name': 'Emerald Pub'}

Specify Fields to Exclude
~~~~~~~~~~~~~~~~~~~~~~~~~

Use the following syntax to specify the fields to exclude from the result:

.. code-block:: java

   projection(fields(exclude("<field name>")))

To use a projection, pass your query filter to the ``find()`` method. Then,
chain the ``projection()`` method to the ``find()`` method.  The following
example uses the ``find()`` method to find all restaurants with
the ``name`` field value of ``"Emerald Pub"``. It then uses a projection to

> exclude the ``grades`` and ``address`` fields from the returned documents.

Reviewer: S: apply same comments as from the include section that I left earlier in the doc

.. io-code-block::
    
   .. input:: /includes/project.java
      :start-after: start-project-exclude
      :end-before: end-project-exclude
      :language: java

   .. output:: 

      {'_id': ObjectId('...'), 'borough': 'Manhattan', 'cuisine': 'American', 'name': 'Emerald Pub', 'restaurant_id': '40367329'}

>       {'_id': ObjectId('...'), 'borough': 'Queens', 'cuisine': 'American',
>       'name': 'Emerald Pub', 'restaurant_id': '40668598'}

Reviewer: S: do not wrap results
```suggestion
      {'_id': ObjectId('...'), 'borough': 'Queens', 'cuisine': 'American', 'name': 'Emerald Pub', 'restaurant_id': '40668598'}
```

When you use a projection to specify which fields to exclude,
any unspecified fields are implicitly included in the return document.

> Troubleshooting
> ---------------

Reviewer: ```suggestion
Projection Errors
-----------------
```

> The following sections describe errors you might see when using projections.

Reviewer: ```suggestion
The following sections describes an error you might encounter when using projections.
```

'Cannot Do Exclusion on Field <field> in Inclusion Projection'

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: S: rename this section to something that summarizes the error
eg. Inclusion and Exclusion Error

> The driver returns an ``OperationFailure`` with this message if you attempt to
> include and exclude fields in a single projection. Ensure that your
> projection specifies only fields to include or fields to exclude.

Reviewer: ```suggestion
The driver returns the following if you attempt to
include and exclude fields in a single projection:

.. code-block:: none

   OperationFailure: ... Cannot Do Exclusion on Field <field> in Inclusion Projection

To resolve this error, ensure that your projection specifies only fields to include or only fields to exclude.
```

read operation:

.. code-block:: java

   projection(fields(include("<field name>")))

To use a projection, pass a query filter to the ``find()`` method. Then, chain
the ``projection()`` method to the ``find()`` method call. 

The following example uses the ``find()`` method to retrieve documents in which
the ``name`` field value is ``"Emerald Pub"``. It then uses a projection to
include only the ``name``, ``cuisine``, and ``borough`` fields from the returned documents.

.. io-code-block::
    
   .. input:: /includes/project.java
      :start-after: start-project-include
      :end-before: end-project-include
      :language: java

   .. output::

      {'_id': ObjectId('...'), 'borough': 'Manhattan', 'cuisine': 'American', 'name': 'Emerald Pub'}
      {'_id': ObjectId('...'), 'borough': 'Queens', 'cuisine': 'American', 'name': 'Emerald Pub'}

When you use a projection to specify fields to include in the return
document, the ``_id`` field is also included by default. All other fields are
implicitly excluded. You must :ref:`explicitly exclude the _id field <java-rs-project-remove-id>`
to omit it from the returned documents.

.. _java-rs-project-remove-id:

> Exclude the ``_id`` Field
> ~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: S: generally, we avoid monospacing in titles
```suggestion
Exclude the _id Field
~~~~~~~~~~~~~~~~~~~~~
```

~~~~~~~~~~~~~~~~~~~~~~~~~

Use the following syntax to specify the fields to exclude from the result of a
read operation:

.. code-block:: java

   projection(fields(exclude("<field name>")))

To use a projection, pass a query filter to the ``find()`` method. Then,
chain the ``projection()`` method to the ``find()`` method call.  

The following Example uses the ``find()`` method to find all restaurants in which
the ``name`` field value is ``"Emerald Pub"``. It then uses a projection to
exclude the ``grades`` and ``address`` fields from the returned documents.

.. io-code-block::
    
   .. input:: /includes/project.java
      :start-after: start-project-exclude
      :end-before: end-project-exclude
      :language: java

   .. output:: 

      {'_id': ObjectId('...'), 'borough': 'Manhattan', 'cuisine': 'American', 'name': 'Emerald Pub', 'restaurant_id': '40367329'}
      {'_id': ObjectId('...'), 'borough': 'Queens', 'cuisine': 'American', 'name': 'Emerald Pub', 'restaurant_id': '40668598'}

When you use a projection to specify which fields to exclude,
any unspecified fields are implicitly included in the return document.

Projection Errors
-----------------

> The following sections describes an error you might encounter when using projections.

Reviewer: ```suggestion
The following section describes an error you might encounter when using projections.
```
